packages/blockchain-wallet-v4-frontend/src/modals/Onboarding/KycVerification/ExtraFields/template.success.tsx

```diff
@@ -135,18 +136,218 @@ const Success: React.FC<InjectedFormProps<{}, Props> & Props> = (props) => {
+      )
 
-  const onChangeNatureAndPurpose = (e, value) => {
-    props.formActions.change(KYC_EXTRA_QUESTIONS_FORM, value, value)
+      if (multipleSelection) {
+        const multipleItems = multipleSelection.map((node) =>
+          node.children.some((item) => item.checked)
+        )
+        return multipleItems.every((item) => item)
+      }
+    }
+    return true
   }
-  const onSourceOfFundsChange = (e, value) => {
-    props.formActions.change(KYC_EXTRA_QUESTIONS_FORM, 'q2', value)
+
+  const updateItem = (nodeId: string, childId: string) => {
+    props.formActions.change(KYC_EXTRA_QUESTIONS_FORM, nodeId, childId)
+
+    const { nodes } = props.extraSteps
+    let isChanged = false
+
+    nodes.map(
+      (node) =>
+        node.id === nodeId &&
+        node.children &&
+        node.children.map((child) => {
+          if (child.id === childId && !child.checked) {
+            child.checked = true
+            isChanged = true
+          }
+          return child
+        })
+    )
+    if (isChanged) {
+      props.identityVerificationActions.updateExtraKYCQuestions({ nodes })
+    }
+  }
+
+  const onChangeInput = (e, value) => {
+    const itemId = e.currentTarget.name
+
+    const { nodes } = props.extraSteps
+    const isChanged = false
+
+    nodes.map(
+      (node) =>
+        node.children &&
+        node.children.map(
+          (child) =>
+            child.children &&
+            child.children.map((item) => {
+              if (item.id === itemId && item.input !== value) {
+                item.input = value
+              }
+              return item
+            })
+        )
+    )
+
+    if (isChanged) {
+      props.identityVerificationActions.updateExtraKYCQuestions({ nodes })
+    }
+  }
+
+  const renderCheckBoxBasedQuestion = (node: NodeType, updateItem) => {
+    return (
+      <FormGroup>
+        <QuestionTitle>{node.text}</QuestionTitle>
+
+        <QuestionDescription>{node.instructions}</QuestionDescription>
+
+        {node.children &&
+          node.children.map((child) => {
+            return (
+              <FormItem key={child.id}>
+                <CheckBoxContainer>
+                  <CenterField>
+                    <CheckBoxText>{child.text}</CheckBoxText>
+                  </CenterField>
+                  <CenterField>
+                    <Field
+                      name={child.id}
+                      id={child.id}
+                      value={child.id}
+                      component={CheckBox}
+                      type='checkbox'
+                      onChange={() => updateItem(node.id, child.id)}
+                    />
+                  </CenterField>
+                </CheckBoxContainer>
+              </FormItem>
+            )
+          })}
+      </FormGroup>
+    )
+  }
+
+  const renderDropDownBasedQuestion = (node: NodeType, updateItem) => {
+    const questionElements = getNodeQuestionElements(node)
```

Comment:
this would be a great place to use the useMemo hook so you don't have do all the the mapping in `getNodeQuestionElements` every time this component renders.
```
const questionElements = useMemo(() => getNodeQuestionElements(node), [node])
```